gguf-py/examples/modify_gguf.py

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python3
+import sys
+from pathlib import Path
+
+# Necessary to load the local gguf package
+sys.path.insert(0, str(Path(__file__).parent.parent))
+
+from gguf import GGUFReader  # noqa: E402
+
+
+def minimal_example(filename: str) -> None:
+    reader = GGUFReader(filename, 'r+')
+    field = reader.fields['tokenizer.ggml.bos_token_id']
+    if field is None:
+        return
+    part_index = field.data[0]
+    field.parts[part_index][0] = 2 # Set tokenizer.ggml.bos_token_id to 2
+    #
+    # So what's this field.data thing? It's helpful because field.parts contains
+    # _every_ part of the GGUF field. For example, tokenizer.ggml.bos_token_id consists
+    # of:
+    #
+    #  Part index 0: Key length (27)
+    #  Part index 1: Key data ("tokenizer.ggml.bos_token_id")
+    #  Part index 2: Field type (4, the id for GGUFValueType.UINT32)
+    #  Part index 3: Field value
+    #
+    # Note also that each part is an NDArray slice, so even a part that
+    # is only a single value like the key length will be a NDArray of
+    # the key length type (numpy.uint32).
+    #
+    # The .data attribute in the Field is a list of relevant part indexes
+    # and doesn't contain internal GGUF details like the key length part.
+    # In this case, .data will be [3] - just the part index of the
+    # field value itself.
+
+
+def change_gguf(reader: GGUFReader, key: str, value: str) -> None:
+    field = reader.get_field(key)
+    if field is None:
+        print(f'! Field {repr(key)} not found', file = sys.stderr)
+        sys.exit(1)
+    # Note that field.types is a list of types. This is because the GGUF
+    # format supports arrays. For example, an array of UINT32 would
+    # look like [GGUFValueType.ARRAY, GGUFValueType.UINT32]
+    handler = reader.gguf_scalar_to_np.get(field.types[0]) if field.types else None
+    if handler is None:
+        print(f'! Field {repr(key)} has unsupported type: {field.types}')
+        sys.exit(1)
+    current_value = field.parts[field.data[0]][0]
+    new_value = handler(value)
+    print(f'* Preparing to change field {repr(key)} from {current_value} to {new_value}')
+    if current_value == new_value:
+        print(f'- Key {repr(key)} already set to requested value {current_value}')
+        sys.exit(0)
+    print('*** Warning *** Warning *** Warning **')
+    print('* Changing fields in a GGUF file can damage it. If you are positive then type YES:')
```

Comment:
I'm quite suspicious of this workflow. We shouldn't allow users to make their file unusable. Instead of a warning, we can check if the new value has the same byte length as the current value. If not, we simply should not make that change and advice creating the model from scratch if possible. Otherwise, I'm almost certain that we'll get multiple issues complaining about it.

Reply:
That script only allows changing simple scalar values, so nothing that can result in a different length. Assuming nothing goes wrong while it's running, it won't violate the GGUF format but changing the value of GGUF fields without knowing what you're doing can definitely make the model unusable.

Reply:
Can this be resolved, or do you think there's still an issue?

Reply:
Looks good, missed the the usage of `scalar_to_np`. Still the message "unsupported type" can be extended but not a big deal as we don't continue.

Reply:
Something like "This tool only supports changing simple values, cannot change field BLAH of type TYPE"?

These tools/examples can, of course, have their functionality expanded in the future. This pull is already pretty big though, so I don't want to throw too much stuff into it.

Doing something like making a general purpose tool to inspect/change fields (including stuff like array fields) wouldn't be that difficult.

Reply:
> Something like "This tool only supports changing simple values, cannot change field BLAH of type TYPE"

Looks good --making sure that users know what to expect is enough.

> Maybe we shouldn't make it too easy to mess with GGUF files though

This was the main reason why I started this conversation. You know, this Python project is a community-maintained one, and the very main reason of its existence to write GGUF files from scratch. So making it too easy to mess up with GGUF files may bring more problems than it can solve. I agree that it can be further developed by time but believe that they should happen bit-by-bit due to common requirements and with good testing.

Reply:
> I agree that it can be further developed by time but believe that they should happen bit-by-bit due to common requirements and with good testing.

Sure. I'm not really planning to expand those scripts in this pull, and if that ever happens in the future then there would have to be some kind of consensus for that pull to get merged.

I think the current approach is reasonable though: you can only use it to change pretty simple values, it prints out what the original value was (so the user can just change it back) _and_ it warns them they should have a backup.

I believe in taking some steps to prevent people from shooting themselves in the foot, but personally I really, really, _really_ hate it when software basically tells me "*Sorry, you can't be trusted to make this decision: I won't let you do that.*"